Added AI Wire Component documentation (#3823)
eclipse/kura: _posts/wires/2022-02-28-kura-wires-ai.md

---
layout: page
title:  "AI Component in Kura Wires"
categories: [wires]
---

The component allows interacting with an `InferenceEngineService` to perform machine learning-related operations.

In a normal machine learning flow, the input is preprocessed before it is given to the inference algorithm, and the result is processed again to be adapted to the rest of the pipeline.

> An `InferenceEngineService` works based on models that describe the inputs and outputs of the algorithm, as well as the algorithm itself. Once these models are loaded in the engine, the AI wire component allows to specify the name of the models that are used in the **pre-processing**, **infer**, and **post-processing** steps. Only the **infer** model name is mandatory so that it is possible to just use the strictly necessary steps.

Reviewer: ```suggestion
An `InferenceEngineService` is a Kura service that implements a simple API to interface with an Inference Engine. The Inference Engine allows to perform inference on trained Artificial Intelligence models commonly described by a file and some configuration for explaining its input and outputs. An example of Inference Engine implementation is the [Triton Inference Server](https://developer.nvidia.com/nvidia-triton-inference-server). Once these models are loaded in the engine, the AI wire component allows to specify the name of the models that are used in the **pre-processing**, **infer**, and **post-processing** steps. Only the **infer** model name is mandatory so that it is possible to just use the strictly necessary steps.
```

> In a normal machine learning flow, the input is preprocessed before it is given to the inference algorithm, and the result is processed again to be adapted to the rest of the pipeline.
> 
> An `InferenceEngineService` is a Kura service that implements a simple API to interface with an Inference Engine. The Inference Engine allows to perform inference on trained Artificial Intelligence models commonly described by a file and some configuration for explaining its input and outputs. An example of Inference Engine implementation is the [Triton Inference Server](https://developer.nvidia.com/nvidia-triton-inference-server). Once these models are loaded in the engine, the AI wire component allows to specify the name of the models that are used in the **pre-processing**, **infer**, and **post-processing** steps. Only the **infer** model name is mandatory so that it is possible to just use the strictly necessary steps.

Reviewer: ```suggestion
An `InferenceEngineService` is a Kura service that implements a simple API to interface with an Inference Engine. The Inference Engine allows to perform inference on trained Artificial Intelligence models commonly described by a file and some configuration for explaining its input and outputs. An example of Inference Engine implementation is the [Triton Inference Server](https://developer.nvidia.com/nvidia-triton-inference-server). 

In a normal machine learning flow, the input is preprocessed before it is given to the machine learning algorithm, and the result is processed again to be adapted to the rest of the pipeline.

Once these models are loaded in the engine, the AI wire component allows to specify the name of the models that are used in the **pre-processing**, **infer**, and **post-processing** steps. Only the **infer** model name is mandatory so that it is possible to just use the strictly necessary steps in case the pre/post-processing is performed directly by the **infer** step.
```

![ai_wires]({{ site.baseurl }}/assets/images/wires/AIComponentConfig.png)

## Models Input and Output formats

The AI wire component takes a `WireEnvelope` as an input, it processes its records and feeds them to the specified preprocessing or inference model. The outputs of the inference or the post-processing step are then reconverted into a wire record. This section explains the inputs and output formats that the wire component is expecting. Not specifying the models according to this contract will result in a non-functioning inference.

The 3 inference steps are applied on each `WireRecord` contained in the input `WireEnvelope`.

> The inputs and outputs will have assigned the corresponding Kura `DataType`, which can be one of:

Reviewer: ```suggestion
The inputs and outputs will be assigned to the corresponding Kura `DataType`, which can be one of:
```

- `BOOLEAN`
- `DOUBLE`
- `FLOAT`
- `INTEGER`
- `LONG`
- `STRING`
-  `BYTE_ARRAY`

Reference to [Wires Introduction](./kura-wires-intro.html) for the data types that are allowed to flow through the wires.

The models that manage the input and the output must expect a list of inputs such that:
- each input corresponds to an entry of the `WireRecord` properties
- the entry key will become the input name (e.g. in the case of an asset, the channel name becomes the tensor name)
- input shape will be `[1]`

In the following, two example configurations for Triton Inference Engine models are provided. A complete usage example can be found in the [Kura GitHub examples repository](https://github.com/eclipse/kura/tree/develop/kura/examples) where the AI component is used to implement an *Anomaly Detector* using the RaspberryPi SenseHat&trade;.

### Input Specification Example

> Following, an example of a model configuration for the Nvidia™ Triton Inference Engine. It expects the input from the `WireEnvelope` that contains a record with properties:

Reviewer: ```suggestion
Following, an example of a [model configuration for the Nvidia™ Triton Inference Engine](https://github.com/triton-inference-server/server/blob/main/docs/model_configuration.md). It expects the input from the `WireEnvelope` that contains a record with properties:
```
